src/core/device.c

```diff
@@ -735,37 +741,105 @@ static bool device_is_ready(sd_device *dev) {
         return r != 0;
 }
 
-static int device_setup_devlink_unit_one(Manager *m, const char *devlink, sd_device **ret) {
-        _cleanup_(sd_device_unrefp) sd_device *dev = NULL;
+static int device_setup_alias_units(Manager *m, sd_device *dev, Set **ready_units, Set **not_ready_units) {
+        _cleanup_strv_free_ char **aliases = NULL;
+        const char *syspath;
+        Device *l;
         int r;
 
         assert(m);
-        assert(devlink);
+        assert(dev);
+        assert(ready_units);
+        assert(not_ready_units);
 
-        if (sd_device_new_from_devname(&dev, devlink) < 0 || !device_is_ready(dev)) {
-                /* the devlink is already removed or not ready */
-                device_update_found_by_name(m, devlink, DEVICE_NOT_FOUND, DEVICE_FOUND_UDEV);
-                *ret = NULL;
-                return 0; /* not ready */
-        }
+        /* On remove, all alias units will be removed by device_update_found_by_sysfs() in device_dispatch_io(). */
+        if (device_for_action(dev, SD_DEVICE_REMOVE))
+                return 0;
 
-        r = device_setup_unit(m, dev, devlink, /* main = */ false);
+        r = sd_device_get_syspath(dev, &syspath);
         if (r < 0)
-                return log_device_warning_errno(dev, r, "Failed to setup unit for '%s': %m", devlink);
+                return r;
+
+        if (device_is_ready(dev)) {
+                const char *s;
+
+                r = sd_device_get_property_value(dev, "SYSTEMD_ALIAS", &s);
+                if (r < 0 && r != -ENOENT)
+                        log_device_warning_errno(dev, r, "Failed to get SYSTEMD_ALIAS property, ignoring: %m");
+                if (r >= 0) {
+                        r = strv_split_full(&aliases, s, NULL, EXTRACT_UNQUOTE);
+                        if (r < 0)
+                                log_device_warning_errno(dev, r, "Failed to parse SYSTEMD_ALIAS property, ignoring: %m");
+                }
+        }
+
+        STRV_FOREACH(alias, aliases) {
+                if (!path_is_absolute(*alias)) {
+                        log_device_warning(dev, "The alias \"%s\" specified in SYSTEMD_ALIAS is not an absolute path, ignoring.", *alias);
+                        continue;
+                }
+
+                if (!path_is_normalized(*alias)) {
+                        log_device_warning(dev, "The alias \"%s\" specified in SYSTEMD_ALIAS is not a normalized path, ignoring.", *alias);
+                        continue;
+                }
+
+                (void) device_setup_unit(m, dev, *alias, /* main = */ false, ready_units);
+        }
+
+        l = hashmap_get(m->devices_by_sysfs, syspath);
+        LIST_FOREACH(same_sysfs, d, l) {
+                if (!d->path)
+                        continue;
+
+                if (path_startswith(d->path, "/dev/"))
+                        continue;
```

Comment:
Is this supposed to skip devlinks? If not, how do you make sure this loop skips devlinks and the same loop in device_setup_devlink_units() skips aliases? Because wouldn't you otherwise maybe end up adding devlinks to `not_ready_units` here and then add them to `ready_units` in the other loop again.

Should we have a sanity check somewhere that there's no unit in both `ready_units` and in `not_ready_units`?

Reply:
OK. Done. Please see the last three commits for core/device.c.